dulwich/config.py

```diff
@@ -252,35 +252,52 @@ def has_section(self, name: Tuple[bytes, ...]) -> bool:
         return name in self.sections()
 
 
-class ConfigDict(Config, MutableMapping):
+BytesLike = Union[bytes, str]
+Key = Tuple[bytes, ...]
+KeyLike = Union[BytesLike, str, Tuple[BytesLike, ...]]
+Value = Union[bytes, bool]
+ValueLike = Union[BytesLike, bool]
+
+
+class ConfigDict(Config, MutableMapping[Key, MutableMapping[BytesLike, Value]]):
     """Git configuration stored in a dictionary."""
 
-    def __init__(self, values=None, encoding=None):
+    def __init__(
+        self,
+        values: Union[
+            MutableMapping[KeyLike, MutableMapping[BytesLike, Value]], None
```

Comment:
I suspect it's desirable for this to be `ValueLike` rather than `Value`; but I don't see anything that uses `self.encoding` to convert the values here to `bytes` so I've typed it as I see it.